Tiloaria/assignment4/Islands.cpp

```diff
@@ -0,0 +1,97 @@
+#include <cstdio>
+#include <cstdlib>
+#include <iostream>
+#include <set>
+#include <vector>
+#include <string>
+#include <assert.h>
+#include <algorithm>
+
+using namespace std;
+
+/*Set value false for all tiles which belong current island.
+Used DFS algorithm.
+In this function map is not empty*/
+void clear_this_island(vector<vector<bool> > &helper_map, int w, int h) {
+    int width = helper_map[0].size();
+    int height = helper_map.size();
+    int add[4][2];//array to go to the next part of island
+    add[0][0] = -1; add[0][1] = 0;
+    add[1][0] = 0; add[1][1] = 1;
+    add[2][0] = 1; add[2][1] = 0;
+    add[3][0] = 0; add[3][1] = -1;
+    helper_map[h][w] = false;
+    for (int i = 0; i < 4; i++) {
+        int new_w = w + add[i][0];
+        int new_h = h + add[i][1];
+        if (new_w < 0 || new_w >= width)
+            continue;
+        if (new_h < 0 || new_h >= height)
+            continue;
+        if (helper_map[new_h][new_w]) {
+            clear_this_island(helper_map, w + add[i][0], h + add[i][1]);
+        }
+    }
+}
+
+/*Count number of islands by map*/
+int count_islands(vector<vector<bool>> map) {
+    int width = 0, height = 0;
+    height = map.size();
+    if (height != 0)
+            width = map[0].size();
```

Comment:
I'm not sure this is a safe assumption to make.. C++ vectors of vectors do not guarantee that map[0].size() == map[n].size(). It's always best to do some error handling to protect your functions against cases like these. In this function you can iterate over the whole vector to make sure all the rows have the same size.

While this might seem counter-intuitive because you're adding a O(N*M) operation when you know the arrays you're passing into this function are rectangular. But it's good to show that you are thinking of possible edge cases and how your code might fail. As an interviewer, this is one of the things I'm actively looking for in interviews.